Hide Payment Request Buttons when cart contains multiple packages

reykjalin opened the pull request:
# Changes proposed in this Pull Request:

Fixes #1504 

When an order is split across multiple packages customers are allowed to choose individual shipping options for each package. This quickly grows in complexity according to `<number_of_packages>^<number_of_shipping_options>` — having just 2 packages with 3 shipping options each results in `2^3 = 8` combinations of shipping options.

We can think of no good way to present this in the Payment Request Dialog that isn't confusing so we decided to hide the buttons when this happens.

# Testing instructions

We need to test that you can pay with PRBs for the following types of orders:

- Contain any amount of products without a shipping package associated with them.
- Contains **only 1** product with a shipping package associated with it.

These types of orders should hide the PRBs:

- Contain **2 or more** products with a shipping package associated with them.

To test this:

1. Make sure you have 2-3 different shipping methods available.
1. Install the [Advanced Shipping Packages plugin](https://woocommerce.com/products/woocommerce-advanced-shipping-packages/).
1. Go to **WooCommerce > Settings > Shipping > Shipping Classes** and create 3 different shipping classes.
1. Go to **WooCommerce > Settings > Shipping > Packages** and create 3 packages, each corresponding to a shipping class.
1. Go to **Products > All Products**, choose 3 different products (e.g. Beanie, Belt, and Sunglasses) and assign each a different shipping class (you'll find this under Product Details > Shipping)
1. Add all 3 products to the cart
1. Go to the Cart page
1. Make sure that no PRBs are visible
    - In `trunk` you should see the PRB
1. Remove any 2 items from the cart.
2. Reload the cart page.
3. See that the PRB is now visible
4. Remove the last item from the cart
5. Add 1 or more items to the cart that don't have a shipping package associated with them.
6. Go the cart page.
7. See that the PRB is visible.

**Make sure to test both the shortcode and block-based carts and checkouts**

Finally, test the single product page; make sure that you can use the PRB no matter whether the product has a shipping package associated with it or not.

-------------------
- [x] Make sure your changes respect [WordPress' coding standards](https://make.wordpress.org/core/handbook/best-practices/coding-standards/).
- [x] Did you make changes, or create a **new .js file**? If **Gruntfile.js** exists in the repo, make sure to run `grunt`.

Commit:
Hide PRBs when cart contains multiple packages

This prevents issues where we have to manually construct shipping
options based on those packages, e.g. 'Free Shipping, Free Shipping,
Priority Shipping' for some 3 products that will be split across 3
packages.

See discussion in #1504 for reference.

ricardo commented:
@reykjalin I found an issue which is not related to this PR, but sort of affects it. We might consider fixing it here since the solution is quite small, although I didn't test it thoroughly.

If you have more than one product with a shipping class added to the cart (where the PRB shouldn't be displayed), and if you go to a simple product page, the PRB is not displayed.

Because of [this line](https://github.com/woocommerce/woocommerce-gateway-stripe/blob/9ae4c809b579f8f4fc56133354ad1d33e6e72ee1/includes/payment-methods/class-wc-stripe-payment-request.php#L616), which should check if this is the cart or checkout page before returning.

Something like this could work, but perhaps you can improve it.

```php
elseif ( ( is_checkout() || is_cart() || $this->is_block() ) && ! $this->should_show_payment_button_on_cart() )
```

ricardo commented on client/blocks/payment-request/index.js:
```diff
@@ -39,6 +39,12 @@ const paymentRequestPaymentMethod = {
 			return false;
 		}
 
+		const shouldShowPaymentRequestButton =
+			getStripeServerData()?.shouldShowPaymentRequestButton ?? false;
+		if ( ! shouldShowPaymentRequestButton ) {
+			return false;
+		}
+
```
Could we simplify this a little bit?

```JavaScript
if ( ! getStripeServerData()?.shouldShowPaymentRequestButton ) {
	return false;
}
```

reykjalin replied:
Nice catch! Fixed in 25682416b6deac74ba534fc2841082c9cc975af9

I had a bit of trouble trying to make this read a bit better, but you're right; your suggestion is much better 😅

ricardo commented on includes/class-wc-stripe-blocks-support.php:
```diff
@@ -81,25 +81,27 @@ public function get_payment_method_script_handles() {
 	 * @return array
 	 */
 	public function get_payment_method_data() {
+		$payment_request_configuration = new WC_Stripe_Payment_Request();
```
Are there any downsides in initializing `WC_Stripe_Payment_Request` here too? Like adding actions/filters twice.

Edit: I added a log to the construct and it looks like this class is being instantiated several times actually 🤔

reykjalin replied:
This will be handled once https://github.com/woocommerce/woocommerce-gateway-stripe/pull/1585 is merged so I didn't address it here.

I should've mentioned it here though, sorry about that!

reykjalin replied:
> class is being instantiated several times actually 🤔

That's weird. I thought this would only be the second time 🤨  I'm not sure where else we're instantiating the class 🤔

ricardo replied:
> I'm not sure where else we're instantiating the class

Well I ran some more tests and apparently when you _refresh_ the **WooCommerce > Status > Logs** page it ends up logging 5 times whatever is in the construct of `WC_Stripe_Payment_Request` 🤷🏻‍♂️. Probably due to multiple redirects + the fact that `WC_Stripe` is initialized with WooCommerce.

When loading the cart page it seems to be initializing it correctly, but twice on the block cart due to the line I commented above.

Since this will be handled in #1585. I'm marking this as resolved 👍

ricardo reviewed (changes requested):
This tests very well 👏 

Code looks great. I just had a few questions.

Commit:
simplify if statement

reykjalin commented:
> I found an issue which is not related to this PR, but sort of affects it. We might consider fixing it here since the solution is quite small, although I didn't test it thoroughly.
>
> If you have more than one product with a shipping class added to the cart (where the PRB shouldn't be displayed), and if you go to a simple product page, the PRB is not displayed.

Nice catch! I like your fix, so I've added that in 0978b203a16f69406652d7f1465cb9720e7fc3fb. I don't think there's a better way to check for that unless we add the conditions to `should_show_payment_button_on_cart()`, and it doesn't feel like they really fit there. Seems like a bit of an unwanted side-effect if we put that there, IMO. I could be convinced otherwise though 🤔

ricardo commented on client/blocks/payment-request/index.js:
```diff
@@ -39,6 +39,10 @@ const paymentRequestPaymentMethod = {
 			return false;
 		}
 
+		if ( ! getStripeServerData()?.shouldShowPaymentRequestButton ) {
+			return false;
+		}
+
```
Sorry, I did not notice this before, but this doesn't look necessary 😅.

The PRB won't show up on the block cart/checkout because `wc_stripe_payment_request_params` won't be available due to [this line](https://github.com/woocommerce/woocommerce-gateway-stripe/blob/0978b203a16f69406652d7f1465cb9720e7fc3fb/includes/payment-methods/class-wc-stripe-payment-request.php#L623).

We can revert the changes in `includes/class-wc-stripe-blocks-support.php` and revert `should_show_payment_button_on_cart` back to `private`.

What do you think?

reykjalin replied:
Hmmm, I think this will ultimately be a better approach considering the work being done in #1585; I'd make `should_show_payment_button_on_cart()` public there anyway so we don't need to duplicate that functionality when we load the [configuration specific to the Blocks checkout](https://github.com/woocommerce/woocommerce-gateway-stripe/pull/1585/files#diff-3197fd215f7f836c9060fc595ce3eef6b21515846bb890f8d3e0a273b605ce28R96-R114) (code from #1585).

You're probably right in that I shouldn't have used this approach in this PR, but it lines up nicely with the changes coming in #1585 so I'd rather keep the change than revert it here just to re-implement it in #1585 😬 

Sorry for the hassle here 🙏 
Let me know what you think!

ricardo replied:
Understood 👍.

It's probably wise to do some of [these checks](https://github.com/woocommerce/woocommerce-gateway-stripe/blob/ed27e97a734d59171aac0c469dfc861533fba865/includes/payment-methods/class-wc-stripe-payment-request.php#L638-L656) as well to determine if the PRB should be displayed on the block cart. You could perhaps add that to #1585.

Maybe considering a generic `should_show_payment_request_button` function [like in WCPay](https://github.com/Automattic/woocommerce-payments/blob/1d678c40e21c1cb16ba288d2d49d8c12f81cf36b/includes/class-wc-payments-payment-request-button-handler.php#L602).

reykjalin replied:
> It's probably wise to do some of these checks as well to determine if the PRB should be displayed on the block cart.

Yeah that's a good idea. I tried adapting what was already in Stripe, but I think you're right; we should refactor this more 🤔 

> You could perhaps add that to #1585 

I'm going to see if I can make this refactor work in this PR 🤔
It seems more related to this PR than it does #1585.

> Maybe considering a generic `should_show_payment_request_button` function like in WCPay.

I think that's what the current function is _sort of_ supposed to be, but it's not quite there. I'll go for that when I refactor this 🙂 

For now I'll label this PR as `in progress` again. Thanks for the reviews Ricardo! 🙏

ricardo replied:
> It seems more related to this PR than it does #1585.

Alright. I was about to submit an approval review 😅, but I'll wait for the next iteration then.

Thanks.

reykjalin replied:
@ricardo I've refactored the `should_show_payment_request_button` function calls significantly and rebased the branch on the latest release. It'd be fantastic if you could take nother look 🙏

Commit:
combine the functions used to check if PRBs should be shown

Commit:
bring more generic checks into should_show_prbs()

This includes checks like whether SSL is enabled, the account keys are
properly configured, etc.

Commit:
add changelog entries

Commit:
bump version numbers in doc comments

ricardo commented on includes/class-wc-stripe-blocks-support.php:
```diff
@@ -100,19 +100,55 @@ public function get_payment_method_data() {
+		//       to version 5.0.
+		if ( function_exists( 'has_block' ) ) {
+			// Don't show if PRBs are supposed to be hidden on the cart page.
+			if (
+				has_block( 'woocommerce/cart' )
+				&& ! apply_filters( 'wc_stripe_show_payment_request_on_cart', true )
+			) {
+				return false;
+			}
```
There's a filter for the checkout page too - `wc_stripe_show_payment_request_on_checkout`, we need to account for that.

```php
if (
    has_block( 'woocommerce/checkout' )
    && ! apply_filters( 'wc_stripe_show_payment_request_on_checkout', false )
) {
    return false;
}
```

reykjalin replied:
Fixed in e2b566ae335893ba78ce3f8b272d532603d5d381.

I made the filter show the PRB on checkout by default _for the checkout block_ because I know that's how the Blocks team wanted the Block to work — it's also how the bundled Stripe integration works in the Blocks plugin.

Let me know if you think we should stick to the same functionality used for the shortcode flow (show on cart, hide on checkout).

ricardo replied:
I'm ok with enabling it to keep the current behavior for blocks, but there's an outstanding issue #1586 which should overlap with this. Basically we will copy the same approach from WCPay and have a setting to display it at specific locations (product, cart and checkout), and remove the filter. This should hide or display the PRB for both checkout block and shortcode. For consistency.

So I don't know what's best to do, TBH. Let me know what you think.

reykjalin replied:
Let's go with the current implementation for now, just to be consistent with what our blocks integration is replacing 🤔 

We can be more consistent when the configuration in #1586 enables a more user friendly we to control this 🙂

ricardo commented on includes/payment-methods/class-wc-stripe-payment-request.php:
```diff
@@ -688,28 +686,42 @@ public function scripts() {
+	 * Returns true if the current page supports Payment Request Buttons, false otherwise.
+	 *
+	 * @since   5.3.0
+	 * @version 5.3.0
```
Update versions to `x.x.x` or next release version here and in other updated methods 🙏

reykjalin replied:
Fixed in ada183454e61b4a4c6a753c17c0427d4cc163892

ricardo commented on includes/payment-methods/class-wc-stripe-payment-request.php:
```diff
@@ -521,6 +521,12 @@ public function allowed_items_in_cart() {
 			return false;
 		}
 
+		// If the cart is not available we don't have any unsupported prodcuts in the cart, so we
```
🤏 Tiny typo in `prodcuts`.

reykjalin replied:
Fixed in 5d008e26a8c0963e2da949b1a93ff36289e7c879

ricardo commented on changelog.txt:
```diff
@@ -1,5 +1,8 @@
 *** Changelog ***
 
+= 5.3.0 - 2021-xx-xx =
+* Fix - Disabled Payment Request Buttons when order has to be split into multiple packages because Payment Requests do not support that use case.
```
Should this read `Disable Payment Request Buttons...`?

reykjalin replied:
Fixed in 517252fa6cc50b1b47ad6e9932958b52796b158f

ricardo commented on includes/payment-methods/class-wc-stripe-payment-request.php:
```diff
@@ -688,28 +686,42 @@ public function scripts() {
+	 */
+	private function is_page_supported() {
+		return $this->is_product()
+			|| is_checkout()
+			|| is_cart()
+			|| WC_Stripe_Helper::has_cart_or_checkout_shortcode_on_current_page()
```
After rebase this will be redundant because `has_cart_or_checkout_on_current_page` checks for `is_cart` and `is_checkout` now due to the latest change for the patch release.

reykjalin replied:
Nice catch! Fixed in 1f0eb7a7f342658d1efeb44a67fd13bd4f2a8d32

ricardo commented on includes/payment-methods/class-wc-stripe-payment-request.php:
```diff
@@ -755,50 +767,79 @@ public function display_payment_request_button_separator_html() {
+		// Don't show if on the cart or checkout page, or if page contains the cart or checkout
+		// shortcodes, with items in the cart that aren't supported.
+		if (
+			( is_cart() || is_checkout() || WC_Stripe_Helper::has_cart_or_checkout_shortcode_on_current_page() )
```
Note: This will be redundant too after rebase. We can either remove `has_cart_or_checkout_on_current_page` from trunk and always use `is_cart` & `is_checkout` explicitly or the other way around.

reykjalin replied:
Fixed in 1f0eb7a7f342658d1efeb44a67fd13bd4f2a8d32

ricardo reviewed (changes requested):
@reykjalin 

This tests well 👍 💯 

Just need to rebase and take into account a few small details as commented below.

Commit:
change updated version numbers to x.x.x

Commit:
fix typo in comment

Commit:
fix typo in changelog

Commit:
fix use of `has_cart_or_checkout` helper function

Commit:
add missing prb display filter to blocks support

ricardo commented:
@reykjalin I noticed when adding a simple product without a shipping package + a product with a shipping package and the quantity is > 1, the payment request doesn't load the shipping options after an address is selected in the PR dialog.

Maybe the solution is to hide duplicate shipping options in `get_shipping_options`? Let me know if you think the change is small enough so we can fix in this PR or if this should be fixed in a separate PR. The changes here LGTM.

ricardo reviewed (approved):
LGTM 👍 

Should we duplicate the changelog entry in readme.txt?

Please check my [previous comment](https://github.com/woocommerce/woocommerce-gateway-stripe/pull/1591#issuecomment-861927697). If you think we can fix that here, I can give it another review, otherwise feel free to merge and we can open an issue.

Commit:
Throw an error when shipping options contain duplicates

This only happens during AJAX calls under very specific circumstances.

reykjalin commented:
> Maybe the solution is to hide duplicate shipping options in get_shipping_options? Let me know if you think the change is small enough so we can fix in this PR or if this should be fixed in a separate PR.

Like we discussed in Slack the root of this problem is probably with `get_shipping_options()`; there's probably something there we need to fix.

I've pushed a commit (12de225539164895b752d6808b65953c06dab1e7) that throws an error when there are duplicate shipping options, just so the Payment Request dialog doesn't break. We might not want it in this PR though, so let me know what you think about that.

I'd consider fixing the `get_shipping_options()` function out of scope for this PR, so we should probably take a look at that another time 🤔

reykjalin commented:
> Should we duplicate the changelog entry in readme.txt?

Yes, I'll do that 😅

Commit:
duplciate changelog in readme

Commit:
fix comment

ricardo reviewed (approved):
LGTM
